File: Engine.cs
using System;
using System.IO;
using System.Collections.Generic;

namespace k5ktool
{
    public class Engine
    {
        const int PATCHES = 128;  // the max number of patches in a bank

        const int SOURCES = 6;  // the max number of sources in a patch

        const int POOL_SIZE = 0x20000;
        const int TONE_COMMON_DATA_SIZE = 82;
        const int NAME_OFFSET = 40;
        const int NAME_SIZE = 8;
        const int SRC_NMBR_OFFSET = 51;

        const int SOURCE_DATA_SIZE = 86;
        const int WAVE_KIT_MSB_OFFSET = 28;
        const int WAVE_KIT_ADD_MASK = 0x04;
        const int ADD_WAVE_KIT_SIZE = 806;

        public Bank ReadBank(string fileName)
        {
            Bank bank = new Bank();
            bank.Patches = new Patch[PATCHES + 1];
            bank.SortedTonePointer = new Patch[PATCHES + 1];
            bank.SortedIndex = new Patch[PATCHES];
            bank.SortedName = new Patch[PATCHES];
            bank.SortedSources = new Patch[PATCHES];
            bank.SortedSize = new Patch[PATCHES];
            bank.SortedCode = new Patch[PATCHES];

            using (FileStream fs = File.OpenRead(fileName))
            {
                Console.WriteLine($"Reading from '{fileName}'");
                using (BinaryReader binaryReader = new BinaryReader(fs))
                {
                    var count = 0;
                    for (int index = 0; index < PATCHES; index++)
                    {
                        Patch patch = new Patch();
                        patch.Index = index;
                        patch.TonePointer = ReadOffset(binaryReader);
                        patch.IsUsed = (patch.TonePointer != 0);
                        if (patch.IsUsed)
                        {
                            bank.SortedIndex[count] = patch;
                            bank.SortedTonePointer[count] = patch;
                            bank.SortedName[count] = patch;
                            bank.SortedSize[count] = patch;
                            bank.SortedSources[count] = patch;
                            bank.SortedCode[count] = patch;
                            count +=1;
                        }

                        patch.AdditiveKitCount = 0;
                        patch.Sources = new Source[SOURCES];

                        for (int sourceIndex = 0; sourceIndex < SOURCES; sourceIndex++)
                        {
                            var source = new Source();
                            source.AdditiveKitPointer = ReadOffset(binaryReader);
                            source.IsAdditive = (source.AdditiveKitPointer != 0);
                            if (source.IsAdditive)
                            {
                                patch.AdditiveKitCount += 1;
                            }
                            patch.Sources[sourceIndex] = source;
                        }

                        bank.Patches[count] = patch;
                    }

                    var lastPatch = new Patch();
                    lastPatch.Index = PATCHES;
                    lastPatch.TonePointer = ReadOffset(binaryReader);
                    bank.SortedTonePointer[count] = lastPatch;

                    bank.PatchCount = count;

                    bank.DataPool = ReadData(binaryReader);

                    Array.Sort(bank.SortedTonePointer, delegate(Patch p1, Patch p2) {
                        return (int)p1.TonePointer - (int)p2.TonePointer;
                    });
                    bank.Base = bank.SortedTonePointer[0].TonePointer;
                    for (int p = 0; p < bank.PatchCount; p++)
                    {
                        var patch = bank.SortedTonePointer[p];
                        patch.TonePointer -= bank.Base;
                        for (int s = 0; s < SOURCES; s++)
                        {
                            var source = patch.Sources[s];
                            if (source.IsAdditive)
                            {
                                source.AdditiveKitPointer -= bank.Base;
                            }
                        }
                    }

                    bank.Patches[PATCHES].TonePointer -= bank.Base;

                    /* analyze number of sources, size and name: */
                    for (int p = 0; p < bank.PatchCount; p++)
                    {
                        var patch = bank.SortedTonePointer[p];
                        patch.SourceCount = bank.DataPool[patch.TonePointer + SRC_NMBR_OFFSET];
                        patch.Size = TONE_COMMON_DATA_SIZE + SOURCE_DATA_SIZE * patch.SourceCount + ADD_WAVE_KIT_SIZE * patch.AdditiveKitCount;
                        patch.Padding = bank.SortedTonePointer[p + 1].TonePointer - patch.TonePointer - patch.Size;

                        // Next up: patch name
                    }
                }
            }

            return bank;
        }

        uint ReadOffset(BinaryReader br)
        {
            var data = br.ReadBytes(4);
            if (BitConverter.IsLittleEndian)
            {
                Array.Reverse(data);
            }
            return BitConverter.ToUInt32(data, 0);
        }

        byte[] ReadData(BinaryReader br)
        {
            var data = br.ReadBytes(POOL_SIZE);
            return data;  // TODO: need to convert endiannness?
        }
    }
}

File: engine.py
NUM_PATCHES = 128
NUM_SOURCES = 6

POOL_SIZE = 0x20000
SOURCE_COUNT_OFFSET = 51
TONE_COMMON_DATA_SIZE = 82
SOURCE_DATA_SIZE = 86
ADDITIVE_WAVE_KIT_SIZE = 806
NAME_OFFSET = 40
NAME_SIZE = 8

SYSEX_HEADER = [0xF0, 0x40]
SYSEX_FOOTER = 0xF7

class Bank:
    def __init__(self):
        self.patches = []
        self.data = None
        self.base = 0

    def add_patch(self, patch):
        self.patches.append(patch)

    def patch_count(self):
        return len(self.patches)

    def get_patch_data(self, patch_index):
        if self.data == None:
            return None
        patch = self.patches[patch_index]
        patch_start_offset = patch.tone_pointer
        patch_end_offset = patch_start_offset + patch.size
        print('data for patch {0} = [{1} : {2}]'.format(patch_index, patch_start_offset, patch_end_offset))
        return self.data[patch_start_offset : patch_end_offset]

class Patch:
    def __init__(self):
        self.is_used = False
        self.additive_kit_count = 0
        self.sources = [None] * NUM_SOURCES
        self.tone_pointer = 0
        self.source_count = 0
        self.source_types = '-' * NUM_SOURCES
        self.size = 0
        self.patch_name = ''
        self.patch_index = 0
    
    def add_source(self, index, source):
        self.sources[index] = source

    def adjust_pointer(self, displacement: int):
        self.tone_pointer -= displacement

    def __str__(self):
        return 'is_used={0} tone_pointer={1:X}'.format(self.is_used, self.tone_pointer)

class Source:
    def __init__(self, pointer: int):
        self.additive_kit_pointer = pointer
        self.is_additive = self.additive_kit_pointer != 0

    def adjust_pointer(self, displacement: int):
        self.additive_kit_pointer -= displacement

    def __str__(self):
        return 'is_additive={0} additive_kit_pointer={1:X}'.format(self.is_additive, self.additive_kit_pointer)

def read_bank(filename):
    bank = Bank()

    with open(filename, 'rb') as f:
        for patch_index in range(0, NUM_PATCHES):
            patch = Patch()
            tone_pointer = int.from_bytes(f.read(4), byteorder='big')
            patch.tone_pointer = tone_pointer
            patch.is_used = tone_pointer != 0
            patch.patch_index = patch_index + 1
            source_pointers = []
            for source_index in range(0, NUM_SOURCES):
                source_pointer = int.from_bytes(f.read(4), byteorder='big')
                source = Source(pointer=source_pointer)
                if source.is_additive:
                    patch.additive_kit_count += 1
                source_pointers.append(source_pointer)
                patch.add_source(source_index, source)
            if patch.is_used:
                bank.add_patch(patch)

        displacement = int.from_bytes(f.read(4), byteorder='big')

        bank.data = f.read(POOL_SIZE)

        tone_pointers = [p.tone_pointer for p in bank.patches]
        tone_pointers.append(displacement)
        sorted_tone_pointers = sorted(tone_pointers)
        base = sorted_tone_pointers[0]

    for p in range(0, bank.patch_count()):
        patch = bank.patches[p]
        patch.adjust_pointer(base)
        for s in range(0, len(patch.sources)):
            source = patch.sources[s]
            if source.is_additive:
                source.adjust_pointer(base)
    displacement -= base

    for p in range(0, bank.patch_count()):
        patch = bank.patches[p]
        patch.source_count = bank.data[patch.tone_pointer + SOURCE_COUNT_OFFSET]
        patch.size = TONE_COMMON_DATA_SIZE + SOURCE_DATA_SIZE * patch.source_count + ADDITIVE_WAVE_KIT_SIZE * patch.additive_kit_count
        name_offset = patch.tone_pointer + NAME_OFFSET
        name_data = bank.data[name_offset : name_offset + NAME_SIZE]
        patch.patch_name = name_data.decode('utf-8')

    for p in range(0, bank.patch_count()):
        patch = bank.patches[p]
        source_types = ''
        for s in range(0, patch.source_count):
            source = patch.sources[s]
            if source.is_additive:
                source_types += 'A'
            else:
                source_types += 'P'
        while s < NUM_SOURCES:
            source_types += '-'
            s += 1
        patch.source_types = source_types

    return (bank, POOL_SIZE - displacement)

dump_commands = {'one_single': [0x20, 0x00, 0x0A, 0x00, 0x00]}

def sysex_message(command, channel: int, data) -> bytearray:
    msg = bytearray(SYSEX_HEADER)
    msg.append(channel)
    msg.extend(dump_commands[command])
    msg.extend(data)
    msg.append(SYSEX_FOOTER)
    return msg
